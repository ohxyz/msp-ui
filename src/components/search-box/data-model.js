/* Data types only used in the component */

const util = require( '../../helpers/util.js' );

class SearchItem {

    constructor( arg ) {

        this.__content__ = '';
        this.__field__ = '';
        this.__origin__ = arg;

        let type = typeof arg;

        if ( type === 'string'
                || type === 'number'
                || type === 'boolean') {

            this.__content__ = arg.toString();
        }
        else if ( arg === null ) {

            this.__content__ = 'null';
        }
        else if ( util.isObject( arg ) === true || arg instanceof SearchItem === true ) {

            this.__content__ = JSON.stringify( arg );

            Object.assign( this, arg );
        }
    }
}

function makeSearchItemsByItems( items ) {

    if ( Array.isArray( items ) === false ) {

        return [];
    }

    let searchItems = items.map( item => {

        return new SearchItem( item );
    } );

    return searchItems;
}



function makeSearchItemsByFields( items, fields ) {

    let searchItems = [];

    items.map( item => {

        fields.map( fieldName => { 

            if ( item.hasOwnProperty( fieldName ) === false ) {

                return ;
            }

            let searchItem = new SearchItem( item );
            searchItem.__content__ = item[ fieldName ];
            searchItem.__field__ = fieldName;

            searchItems.push( searchItem );

        } );

    } );

    return searchItems;
}

function makeSearchItems( items, fields ) {

    if ( Array.isArray( fields ) === true && fields.length > 0 ) {

        return makeSearchItemsByFields( items, fields );
    }

    return makeSearchItemsByItems( items );

}

module.exports = {

    SearchItem,
    makeSearchItems,
    makeSearchItemsByItems,
    makeSearchItemsByFields,
};