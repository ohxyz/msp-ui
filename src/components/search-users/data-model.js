/* Store and process users data from SAP
 *
 */

class UserStorage {

    constructor( sapRaw ) {

        this.sapData = sapRaw;
        this.users = [];
        this.orgs = [];
    }

}

module.exports = {

    UserStorage,
};