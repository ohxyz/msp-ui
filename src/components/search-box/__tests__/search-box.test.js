import React from 'react';
import SearchBox from '../search-box.js';
import { SearchItem } from '../data-model.js';
import renderer from 'react-test-renderer';

const items = [

    { 'org': 'abcd Education', name: 'abcd' },
    { 'org': 'abce Health', name: 'abce health people' },
    { 'org': 'abcde Transport' },
    { 'name': 'Mr abcdg ' },
    { 'name': 'abcd lady', 'org': 'abce LADY GROUP' }
];

it( 'renders with default settings', () => {

    const tree = renderer
        .create( 
            <SearchBox
                id="seach-box-users"
                name="my-box"
                onSelect={ ( item, self ) => { console.log( 'onSelect', item ); } }
                onIconClick= { self => { console.log( 'onIconClick', self ); self.showAllItems(); } }
                onChange= { self => console.log( 'onChange', self ) }
                placeholder="Search users by name, department or agency"
                onFocus={ self => { console.log( 'focus', self ); } }
                onBlur={ self => console.log( 'blur', self ) }
                items={ items }
                iconStyle="add"
                fields={ [ 'org', 'name' ] }
                strikes={ 3 } /> 
        )
        .toJSON();

    expect( tree ).toMatchSnapshot();

} );


describe( 'SearchBox React component instance', () => {

    // Stub
    let items = [

        'abcd', 'abce', 'bcde', 'cdef', 'cefg', 'defg', 'efgh'
    ];

    describe( 'static method: getDerivedStateFromProps', () => { 

        let state = SearchBox.getDerivedStateFromProps( {}, {} );
        expect( Object.keys( state ).length ).toBe( 18 );

    } );
    
    describe( 'filterSearchItemsByText method', () => {

        let props = { items: items };
        let searchBox = new SearchBox( props );

        test( 'can filter searchItems by text', () => {

            let searchItems = items.map( item => new SearchItem( item ) );

            let filtered = searchBox.filterSearchItemsByText( searchItems, 'abc' );
            let expected = [

                { __content__: 'abcd', __origin__: 'abcd', '__field__': '' },
                { __content__: 'abce', __origin__: 'abce', '__field__': '' },

            ];

            expect( filtered ).toEqual( expected );

        } );

    } );

    describe( 'updateItems method', () => {

        let props = { items: items };
        let searchBox = new SearchBox( props );
        let spy = jest.spyOn( searchBox, 'filterSearchItemsByText' );

        // Mock
        searchBox.textInputElement = { value: 'mocked value' };
        searchBox.setState = new Function();
        searchBox.updateItems();

        expect( spy ).toHaveBeenCalled();

    } );


    describe( 'renderCount method', () => { 

        let searchBox = new SearchBox( {} );
        searchBox.state.shouldRenderCount = true;

        let result = searchBox.renderCount();

        expect( result.props.children[ 0 ].props.children ).toBe( 0 );
    } );

    describe( 'sortByIndexOfFields method', () => {

        

    } );

} );

