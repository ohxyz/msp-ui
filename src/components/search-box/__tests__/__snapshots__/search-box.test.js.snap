// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox creates snapshots 1`] = `
<div
  className="search-box"
  id="seach-box-users"
>
  <div
    className="search-box__header"
  >
    <input
      className="search-box__field"
      data-text=""
      id={undefined}
      name="my-box"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search users by name, department or agency"
      type="text"
    />
    <span
      className="search-box__icon"
      onClick={[Function]}
    >
      <i
        className="material-icons"
      >
        add
      </i>
    </span>
  </div>
</div>
`;
